frappe/model/document.py

```diff
@@ -1250,6 +1253,38 @@ def check_no_back_links_exist(self):
 			check_if_doc_is_linked(self, method="Cancel")
 			check_if_doc_is_dynamically_linked(self, method="Cancel")
 
+	def save_snapshot(self):
+		"""Save snapshot of the document"""
+		if (
+			not self._doc_before_save
+			or self.doctype == "Version"
+			or frappe.flags.in_patch
+			or frappe.flags.in_install
+			or not getattr(self.meta, "allow_document_snapshots", False)
+			or self.docstatus != 0
+		):
+			return
+		old_doc = self._doc_before_save
+		# Remove the name key from the child record, as storing it in snapshot
+		# will remove the child record when restoring the snapshot
+		old_doc = self.delete_name_from_child_table(old_doc)
+		version = frappe.new_doc("Version")
+		version.ref_doctype = self.doctype
+		version.docname = self.name
+		version.complete_snapshot = True
+		version.data = json.dumps(old_doc.as_dict(), default=str)
+		version.insert(ignore_permissions=True)
```

Comment:
So... your design will insert two version docs for change?
1. One for diff 
2. another for complete snapshot. 


Now that I see this as PR I can see future problem related to this implementation and also @safwansamsudeen's impl. 

If we reuse Version table then we don't be able to clean up old versions effectively. 

I also want snapshots to be clearable by using `Log Settings`. `Version` can't be cleared in most countries for compliance reasons but snapshots are fine ig.

Deleting few rows in the middle of `Version` table with some filter is also bad idea from MySQL POV. Let's keep Version immutable only maybe :sweat_smile:

Reply:
It makes more sense to use a seperate `Snapshot` doctype, especially as we don't need built in diffing. 

My impl doesn't use two version docs - it adds an `image` field to `Version` and saves all the data. It's rather slick (like two lines of code change), but if you want snapshots to be clearable, we should use a new doctype.

Reply:
Agree, It makes sense to have a new doctype for it.